### PR TITLE
tools to convert from lcov and gocover to Codecov json

### PR DESCRIPTION
**Pull request by RaduBerinde**
I have already checked in these tools to the master branch but wanted to give you a chance to review. I will apply any updates to master.

### CODE REVIEW

File: gocover2json/main.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	// (instead of being part of an array). This makes it hard to use Go's json
+	// machinery; we just produce the output directly.
+	w := bufio.NewWriter(jsonWriter)
+	w.WriteString("{\n")
+	w.WriteString("  \"coverage\": {")
+	for fileIdx, profile := range profiles {
+		var lines []lineCount
+		for _, b := range profile.Blocks {
+			for len(lines) <= b.EndLine {
+				lines = append(lines, -1)
```
Comment (srosenberg) [inline gocover2json/main.go]: Might be worth a short comment that lines are 1-based, not 0-based. I had to pause for ~15 seconds :)

File: lcov2json/main.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+)
+
+// lcov2json is a program that converts from LCOV format to the Codecov JSON
```
Comment (srosenberg) [inline lcov2json/main.go]: Let's also link LCOV format doc., https://ltp.sourceforge.net/coverage/lcov/geninfo.1.php

File: lcov2json/main.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+		os.Exit(2)
+	}
+}
+
+// lineCount is the hit count for a line, or -1 if the line has no count.
+type lineCount int
+
+const noCount lineCount = -1
+
+func convertLcovToJson(lcovReader io.Reader, jsonWriter io.Writer) error {
```
Comment (srosenberg) [inline lcov2json/main.go:71]: Do we envision this being used outside of CLI?

File: lcov2json/main.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+				w.WriteString("\n")
+				fmt.Fprintf(w, "      \"%d\": %d", i, count)
+			}
+			w.WriteString("\n    }")
+
+			currentFile = ""
+		}
+		idx := strings.Index(l, ":")
+		if idx == -1 {
+			// Don't know how to parse this line; skip.
```
Comment (srosenberg) [inline lcov2json/main.go:114]: Emit a warning?

File: lcov2json/main.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+			currentFile = val
+			currentLines = currentLines[:0]
+		case "DA":
+			var line, count int
+			_, err := fmt.Sscanf(val, "%d,%d", &line, &count)
+			if err != nil {
+				return fmt.Errorf("error parsing DA line: %v", err)
+			}
+			// Sanity check
+			if line > 1000000 {
```
Comment (srosenberg) [inline lcov2json/main.go:131]: Makes me wonder what is the largest source file we have in crdb atm? :)

**Review by srosenberg — APPROVED** [review]
Nice!

File: gocover2json/main.go (hunk 1 above)
Comment (RaduBerinde) [reply gocover2json/main.go]: Renamed to `lineCounts` and added a comment.

File: lcov2json/main.go (hunk 2 above)
Comment (RaduBerinde) [reply lcov2json/main.go]: Done.

File: lcov2json/main.go (hunk 3 above)
Comment (RaduBerinde) [reply lcov2json/main.go:71]: For now at least, no..

File: lcov2json/main.go (hunk 4 above)
Comment (RaduBerinde) [reply lcov2json/main.go:114]: Done.

File: lcov2json/main.go (hunk 5 above)
Comment (RaduBerinde) [reply lcov2json/main.go:131]: I hope it's nowhere close :)

**Comment by RaduBerinde** [comment]
TFTR!

**Comment by RaduBerinde** [comment]
Updates are in master, closing.

**Pull request closed without merging**